### ISSUE TITLE
Enhancement: Gauss expansion compiler pass for complex bilinear ops

### ISSUE THREAD

**Issue by benmacadam64** [enpasos/jax2onnx#127]
The complex number support looks great! I sketched out a solution to get general bilinear operations on complex arrays - I'm sure you have some thoughts of your own, but maybe this can get the ball rolling. 

I think the cleanest approach would be to use a compiler pass from `onnx-ir` to expand unsupported nodes. It should just be a graph transformer pass, where you instantiate the nodes as though they're well-defined and then performa compiler pass for Gauss expansion. Here is a sketch of a reasonably future-proof approach; you'd just need to update the `Real/Imag/PackComplex` nodes whenever more support for complex operations is added.

```
Ar <- Real(A)
Ai <- Imag(A)
Br <- Real(B)
Bi <- Imag(B)

P1 <- Node(Ar, Br)
P2 <- Node(Ai, Bi)
P3 <- Node(Sum(Ar, Ai), Sum(Br, Bi))

C = PackComplex(Sub(P1, P2), Sub(Sub(P3, P1), P2))
```
And you could have a simpler option 4 reals:
```
Ar <- Real(A)
Ai <- Imag(A)
Br <- Real(B)
Bi <- Imag(B)

A-B= <- Node(A-, B-) (for -, = in (r, i))

C = PackComplex(Sub(ArBr,AiBi), Sum(AiBr, ArBi))
```

You may want to have this wrapped in an ONNX function for a clean Netron visualization.

Notes:
- **Einsum:** keep the original equation and pass the real/imag components as independent inputs.  
  Ensure the trailing complex channel (`[..., 2]`) is *not* part of the contraction labels; treat it as a feature dimension preserved through the op.

- **Conv / ConvTranspose:** identical Gauss expansion applies.  
  Insert a `Transpose` only if you need to convert between channel-last packing (`NHWC...2`) and the backend’s preferred NCHW layout.  
  Reuse stride, padding, dilation, and groups attributes verbatim. Packed-real helpers already stamp correct metadata.

- **Conjugation (adjoint):** define a `Conj(X)` ONNX Function implemented as  
  `PackComplex(Real(X), Neg(Imag(X)))`.  
  You can insert this before the Gauss expansion, or equivalently flip the signs of the imaginary terms in the epilogue wiring.

- **Real inputs:** when one operand is real, call `ensure_packed_real_pair(...)` from `_complex_utils.py`;  
  it produces a packed tensor with an explicit zero imaginary channel so the pass can safely set `P2 = 0` and `P3 = Op(Ar, Br)`.

- **Strategy knob:** expose a graph-level attribute `complex_strategy ∈ {"gauss", "four_real"}`.  
  Use the Gauss 3-op form by default; fall back to `four_real` in low-precision (`FLOAT16` / `BFLOAT16`) or numerically sensitive paths.

- **Bias / α / β handling:** push scalar parameters into each `P*` node and add the real / imaginary bias to `(Cr, Ci)` before `PackComplex`.  
  When the bias itself is complex, normalise it through `ensure_packed_real_pair(...)`.

- **Broadcasting and attributes:** preserve all original op attributes and rely on ONNX’s native broadcasting semantics (`MatMul`, `Einsum`, `Conv`, …).  
  Shape and dtype stamping should reuse `coerce_dim_values(...)` and `_stamp_type_and_shape(...)` for consistency.

- **Validation:** compare the lowered graph numerically against NumPy / JAX complex results over batched and broadcasted shapes.  
  Include FP16 / BF16 stress tests and conjugate / transpose variants.

**Comment by enpasos** [comment]
Thanks again for the sketches that kicked this all off—they were super helpful in charting the path forward.

We’ve now merged the first wave of complex-number support into `main`:
- Unified packed real/imag layout `[..., 2]`
- Helper stack for packing/unpacking and dtype reconciliation
- Initial coverage:
  • Elementwise: `lax/jnp` `add` / `sub` / `mul` / `div`  
  • Conjugation: `lax.conj` / `jnp.conj`  
  • Bilinear ops: `dot_general` / `jnp.matmul`, `conv_general_dilated`  
  • Transforms: FFT via ONNX-compliant DFT

**Note on strategy:** at the moment we always lower via the straightforward *four-real* expansion. The `complex_strategy` knob isn’t wired yet; to pick that up later I’ve added a “Potential optimizations” section in `docs/dev_guides/complex_numbers.md` (e.g., the Gauss 3-multiply form).

It’s a solid base that should make it straightforward to extend complex handling to more components. Closing this to keep the tracker tidy—if you see ops that would benefit from complex coverage, please open a focused follow-up or a PR.
